File: hookline/examples/consumers/verify.php
<?php

/**
 * Verify a hookline (Standard Webhooks) signature.
 *
 * $body must be the raw request bytes: file_get_contents('php://input'),
 * not $_POST or a decoded array that has been re-encoded, which is different
 * bytes and will not match.
 *
 * @param array<string,string> $headers lower-cased header names
 * @throws RuntimeException when the request is not authentic
 */
function hookline_verify(
    string $secret,
    array $headers,
    string $body,
    int $toleranceSeconds = 300
): bool {
    $id = $headers['webhook-id'] ?? null;
    $timestamp = $headers['webhook-timestamp'] ?? null;
    $signature = $headers['webhook-signature'] ?? null;
    if ($id === null || $timestamp === null || $signature === null) {
        throw new RuntimeException('missing webhook headers');
    }

    if (abs(time() - (int) $timestamp) > $toleranceSeconds) {
        throw new RuntimeException('stale timestamp');
    }

    $key = base64_decode(preg_replace('/^whsec_/', '', $secret), true);
    if ($key === false) {
        throw new RuntimeException('the secret is not base64');
    }
    $expected = hash_hmac('sha256', "{$id}.{$timestamp}.{$body}", $key, true);

    // Several space-separated signatures during a rotation; one is enough.
    foreach (explode(' ', $signature) as $part) {
        $pieces = explode(',', $part, 2);
        if (count($pieces) !== 2 || $pieces[0] !== 'v1') {
            continue;
        }
        $given = base64_decode($pieces[1], true);
        if ($given !== false && hash_equals($expected, $given)) {
            return true;
        }
    }
    throw new RuntimeException('no signature matched');
}

File: hookline/examples/consumers/verify.go
package webhook

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"net/http"
	"strconv"
	"strings"
	"time"
)

var (
	ErrMissingHeaders = errors.New("missing webhook headers")
	ErrStaleTimestamp = errors.New("stale timestamp")
	ErrNoMatch        = errors.New("no signature matched")
)

// Verify checks a hookline (Standard Webhooks) signature.
//
// body must be the raw request bytes, read with io.ReadAll(r.Body) before
// anything decodes them: a struct that has been unmarshalled and marshalled
// again is different bytes and will not match.
func Verify(secret string, headers http.Header, body []byte, tolerance time.Duration) error {
	id := headers.Get("webhook-id")
	timestamp := headers.Get("webhook-timestamp")
	signature := headers.Get("webhook-signature")
	if id == "" || timestamp == "" || signature == "" {
		return ErrMissingHeaders
	}

	seconds, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return ErrStaleTimestamp
	}
	if age := time.Since(time.Unix(seconds, 0)); age > tolerance || age < -tolerance {
		return ErrStaleTimestamp
	}

	key, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(secret, "whsec_"))
	if err != nil {
		return err
	}
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(id + "." + timestamp + "."))
	mac.Write(body)
	expected := mac.Sum(nil)

	// Several space-separated signatures during a rotation; one is enough.
	for _, part := range strings.Split(signature, " ") {
		version, value, found := strings.Cut(part, ",")
		if !found || version != "v1" {
			continue
		}
		given, err := base64.StdEncoding.DecodeString(value)
		if err != nil {
			continue
		}
		if hmac.Equal(given, expected) {
			return nil
		}
	}
	return ErrNoMatch
}
